### ISSUE TITLE
JavaScript validation

### ISSUE THREAD

**Issue by pedzed** [alexgarrett/violin#53]
The only secure way to validate is by doing it through the back end. That would mean that the server has to send a response: the page, or a part of the page (in case of AJAX), has to be reloaded.

To improve the user experience on websites, you should provide client-side (JavaScript) validation as well. Client side validation will allow developers to do "real time" validation. Every time a form values changes, you can update error messages. For example, you have a text input, and you check for `onkeyup`. If the user releases a key, you could revalidate.

This issue is a feature request to provide a convenient way to pass Violin's settings to JavaScript.

As long as this feature request will not be implemented, developers will have to write their own or use another library for JavaScript validation. This causes duplication of rules and messages, which will significantly increase development time.

**Comment by tleb** [comment]
> This causes duplication of rules and messages

Rules are anyway written in PHP, they will need to be re-written in Javascript.

It would be like asking the [Carbon](https://github.com/briannesbitt/Carbon) creator to adapt his class to Javascript : it would be an entirely different project.

**Comment by pedzed** [comment]
> Rules are anyway written in PHP, *they will need to be re-written in Javascript*.

No, because you can pass data from PHP to JavaScript (e.g. through JSON).

**Comment by HackinGuy** [comment]
Instead of rewriting the Violin in JavaScript there is another solution that I can think of.
Create your own kind of a "protocol" that with it you will define your Violin setting.
For example, send using ajax the content and requirements as json.
The PHP page will understand the request break it into pieces, use Violin to validate and send the result.
This solution is mainly for the simple built in function like email, required, alpha etc..
I can't really think on a simple solution for more complex validations.

**Comment by tleb** [comment]
How to overload your server ? Well, make him do what the client is supposed to do. :smile:

**Comment by HackinGuy** [comment]
@tleb I think that what you say is incorrect, calculations that Violin use are very minor and barely use CPU or memory. The problem with the validation that @pedzed is trying to achieve is that it might be slow and buggy if the server is slow or the user has a very slow internet connection.

**Comment by tleb** [comment]
My point is that validation can really easily be done by the client, and it would be stupid to use the server's resources (not only CPU or memory, but also broadband) to do such a basic task. Of course, the check should also be done server-side, when it receives the information, as the user could have modified the Javascript, but this is only once per query, when the server receives the data.

**Comment by HackinGuy** [comment]
@tleb you are correct. I believe that @pedzed wants to do it in order to save time matching the JS validation and the PHP validation. Anyway another solution that @pedzed can use is to copy all of the rules(regex mainly) from violin and use them in JS if Alex would agree.

**Comment by pedzed** [comment]
> Anyway another solution that @pedzed can use is to copy all of the rules(regex mainly) from violin and use them in JS if Alex would agree.

That is actually exactly what I meant. (Was first paragraph confusing?) I disagree with the regex's though.

The server passes the Violin rules and messages once to the client, so that you can perform client side validation using JavaScript.

The only problem is validation that has to be done server side (e.g. checking the database).
